_overviews/core/programming-collections-generically.md

```diff
@@ -0,0 +1,320 @@
+  new IntersperseOperation(seq(coll))
+~~~
+
+However, if we try it we get the following behaviour:
+
+~~~
+scala> List(1, 2, 3).intersperse(0)
+res0: Array[Int] = Array(1, 0, 2, 0, 3)
+~~~
+
+We get back an `Array` although the source collection was a `List`! Indeed, there is
+nothing that constrains the result type of `interseprse` to depend on the receiver type.
+The following expressions show the behavior we expect:
+
+~~~ scala
+List(1, 2, 3).intersperse(0) == List(1, 0, 2, 0, 3)
+"foo".intersperse(' ') == "f o o"
+~~~
+
+When we call it on a `List`, we want to get back another `List`, and when we call it on
+a `String` we want to get back another `String`, and so on.
+
+To produce a collection whose type depend on a source collection, we have to use
+`scala.collection.BuildFrom` (formerly `CanBuildFrom`) instead of `Factory`.
+`BuildFrom` is defined as follows:
+
+~~~ scala
+trait BuildFrom[-From, -A, +C] {
+  /** @return A collection of type `C` containing the same elements as the source collection `it`. */
+  def fromSpecificIterable(from: From)(it: Iterable[A]): C
+
+  /** Get a Builder for the collection type `C` */
+  def newBuilder(from: From): Builder[A, C]
+}
+~~~
+
+`BuildFrom` has similar operations to `Factory`, but they take an additional `from`
+parameter. Before explaining how implicit instances of `BuildFrom` are resolved, let’s first have
+a look at how you can use it. Here is the implementation of `intersperse` based on `BuildFrom`:
+
+~~~ scala
+import scala.collection.{ AbstractView, BuildFrom }
+import scala.collection.generic.IsSeqLike
+
+class IntersperseOperation[Repr, S <: IsSeqLike[Repr](coll: Repr, seq: S) {
+  def intersperse[B >: seq.A, That](sep: B)(implicit bf: BuildFrom[Repr, B, That]): That = {
+    val seqOps = seq(repr)
+    bf.fromSpecific(coll)(new AbstractView[B] {
+      // same as before
+    })
+  }
+}
+
+implicit def IntersperseOperation[Repr](coll: Repr)(implicit seq: IsSeqLike[Repr]): IntersperseOperation[Repr, seq.type] =
+  new IntersperseOperation(coll)(seq)
+~~~
+
+Note that we track the type of the receiver collection `Repr` in the `IntersperseOperation`
+class. Now, consider what happens when we write the following expression:
+
+~~~ scala
+List(1, 2, 3).intersperse(0)
+~~~
+
+An implicit parameter of type `BuildFrom[Repr, B, That]` has to be resolved by the compiler.
+The type `Repr` is constrained by the receiver type (here, `List[Int]`) and the type `B` is
+inferred by the value passed as a separator (here, `Int`). Finally, the type of the collection
+to produce, `That` is fixed by the resolution of the `BuildFrom` parameter. In our case,
+there is a `BuildFrom[List[Int], Int, List[Int]]` instance that fixes the result type to
+be `List[Int]`.
+
+## Summary
+
+- To consume any collection, take an `IterableOnce` (or something more specific) as parameter,
+  - To also support `String`, `Array` and `View`, use `IsIterableLike`,
```

Comment:
whoa, where did `View` come from? I don't recall that being mentioned before, I only recall `String` and `Array`. that seems important and should probably be worked back into the main text.